dao/rhc_connection_dao.go

```diff
@@ -0,0 +1,255 @@
+		Where(`"jt"."tenant_id" = ?`, s.TenantID).
+		Group(`"rhc_connections"."id"`).
+		Limit(limit).
+		Offset(offset)
+
+	query, err := applyFilters(query, filters)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// Run the actual query.
+	result, err := query.Rows()
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// We call next as otherwise "ScanRows" complains, but since we're going to map the results to an array of
+	// map[string]interface{}, "ScanRows" will already scan every row into that array, thus freeing us from calling
+	// result.Next() again.
+	result.Next()
+
+	// Loop through the rows to map both the connection and its related sources.
+	var rows []map[string]interface{}
+	err = DB.ScanRows(result, &rows)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	rhcConnections := make([]m.RhcConnection, 0)
+	for _, row := range rows {
+		rhcConnection, err := mappers.MapRowToRhcConnection(row)
+		if err != nil {
+			return nil, 0, err
+		}
+
+		rhcConnections = append(rhcConnections, *rhcConnection)
+	}
+
+	err = result.Close()
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// Getting the total count (filters included) for pagination.
+	count := int64(0)
+	query.Count(&count)
+
+	return rhcConnections, count, nil
+}
+
+func (s *RhcConnectionDaoImpl) GetById(id *int64) (*m.RhcConnection, error) {
+	query := DB.
+		Debug().
+		Model(&m.RhcConnection{}).
+		Select(`"rhc_connections".*, STRING_AGG(CAST ("jt"."source_id" AS TEXT), ',') AS "source_ids"`).
+		Joins(`INNER JOIN "source_rhc_connections" AS "jt" ON "rhc_connections"."id" = "jt"."rhc_connection_id"`).
+		Where(`"rhc_connections"."id" = ?`, id).
+		Where(`"jt"."tenant_id" = ?`, s.TenantID).
+		Group(`"rhc_connections"."id"`)
+
+	// Run the actual query.
+	result, err := query.Rows()
+	if err != nil {
+		return nil, err
+	}
+
+	// We call next as otherwise "ScanRows" complains, but since we're going to map the results to an array of
+	// map[string]interface{}, "ScanRows" will already scan every row into that array, thus freeing us from calling
+	// result.Next() again.
+	if !result.Next() {
+		return nil, util.NewErrNotFound("rhcConnection")
+	}
+
+	// Loop through the rows to map both the connection and its related sources.
+	var rows []map[string]interface{}
+	err = DB.ScanRows(result, &rows)
+	if err != nil {
+		return nil, err
+	}
+
+	err = result.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	if len(rows) != 1 {
+		return nil, errors.New("unexpected number of results")
+	}
+
+	rhcConnection, err := mappers.MapRowToRhcConnection(rows[0])
+	if err != nil {
+		return nil, err
+	}
+
+	return rhcConnection, nil
+}
+
+func (s *RhcConnectionDaoImpl) Create(rhcConnection *m.RhcConnection) (*m.RhcConnection, error) {
+	// If the source doesn't exist we cannot create the RhcConnection, since it needs to be linked to at least one
+	// source.
+	var sourceExists bool
+	err := DB.Debug().
+		Model(&m.Source{}).
+		Select(`1`).
+		Where(`id = ?`, rhcConnection.Sources[0].ID).
```

Comment:
Sorry I missed this yesterday - we need a tenancy check otherwise someone could attach to any source!

Reply:
Great catch! Added a test as well to avoid regressions.